File: src/main/java/ting/leetcode/java/arrays101/TwoSum.java
package ting.leetcode.java.arrays101;

import java.util.HashMap;

public class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> differenceToIndex = new HashMap<>();
        int[] result = new int[2];

        for (int i = 0; i < nums.length; i++) {
            int element = nums[i];
            int difference = target - element;
            if (differenceToIndex.containsKey(element)){
                return new int[] {differenceToIndex.get(element), i};
            }
            differenceToIndex.put(difference, i);
        }
        return result;
    }
}

File: kotlin-solution/src/leetcode/hashtable/TwoSum.kt
package leetcode.hashtable

class TwoSum {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val map = mutableMapOf<Int, Int>()
        val result = IntArray(2)
        nums.forEachIndexed { index, value ->
            map[value]?.let {
                result[0] = nums.indexOf(it)
                result[1] = index
            }
            map[target - value] = value
        }
        return result
    }
}
